src/java/org/apache/cassandra/service/ReplicaFilteringProtection.java

```diff
@@ -239,7 +253,21 @@ public Row onMergedRows(Row merged, Row[] versions)
                 {
                     // cache the row versions to be able to regenerate the original row iterator
                     for (int i = 0; i < versions.length; i++)
+                    {
                         builders[i].addRow(versions[i]);
+                        rowsCached++;
+
+                        if (rowsCached == cachedRowsFailThreshold)
+                        {
+                            throw new TooManyCachedRowsException(cachedRowsFailThreshold, command);
+                        }
+                        else if (rowsCached == cachedRowsWarnThreshold)
+                        {
+                            logger.warn("Replica filtering protection has cached {} rows during query {}. " +
```

Comment:
We could issue also a client warning here, as we do with `tombstone_warn_threshold` in [`ReadCommand`](https://github.com/apache/cassandra/blob/48c54e49a636480b07dde5b12fe83cb8abfab2cd/src/java/org/apache/cassandra/db/ReadCommand.java#L543).

Reply:
Agreed. I'll try to graft that in here...